074-additional-volumes-support.md

```diff
@@ -0,0 +1,95 @@
+# Support of additional volumes in pod
+
+We propose to enhance the Strimzi Kafka Operator by adding support for specifying additional volumes in the CRD. This will involve modifying the Kafka CRDs to include an `additionalVolumes` field, which will allow users to define extra volumes and volume mounts.
+
+The specific usecase of ours, is that we want to set custom log4j logging for audit logs, to ensure that audit logs are *not* present in stdout, but can be picked up by a sidecar which uses OpenTelemtry or logstash to forward to the audit logs to a secure location.
+
+## Current situation
+
+The operator handles many configurations and operational aspects but lacks support for specifying additional volumes in the Custom Resource Definition (CRD). This limitation can restrict users from accessing log files or integrating with external storage solutions directly.
+
+## Motivation
+
+Adding support for additional volumes in the CRD will allow users to:
+
+ - Access log files for enhanced monitoring and debugging.
+ - Use additional configurations or secret management tools that require mounting extra volumes.
+
+Improve flexibility in managing Kafka deployments by allowing users to attach custom storage solutions based on their specific requirements.
+
+## Proposal
+
+### Considerations
+
+- CRD Changes: Update the Kafka CRDs to include an `additionalVolumes` field.

```

Comment:
Please be more exact what CRDs and which parts of them will be affected and which operands will be implemented as `Kafka CRDs` is a bit confusing. I would expect it to be implemented for all CRDs, all operands ... including all individual components of them? E.g. the Kafka CR has not just the Kafka node pods. It has also ZooKeeper pods, Cruise Control pods, Entity OPerator Pods, Kafka Exporter etc. Then there are the others such as Connect, Bridge, MirrorMaker etc. So it would be good to understand hat will be changed, where and how.

The simplest answer is all should change. But I can imagine for example saying that ZooKeeper pods or Mirror Maker 1 pods will be ignored as they will be removed soon etc.

Reply:
I will investigate the codebase further. When making changes to the PodTemplate.java the CRDGenerator only modified the CRD for KafkaCluster/brokers. I think we should also consider what would requiress less implementation effort, but is made in way so we can extend it over time to support more types.

But again, I think we need to understand the codebase a bit better before scoping this.

Reply:
I guess it is up for discussion. The first thing is to at least clarify the intention of the proposal - I do not think it is clear right now. There are also various options. For example having a generic proposal covering everything but implement it only somewhere.

My personal view (might differ from other maintainers) is that I think for example the demand for using custom volumes with Connect is equal if not bigger to the Kafka brokers. Plus I think many of the usecases (including yours) more or less apply to all of the components. So I would probably tend to go with doing all of them.

I did not check your PR. But the [PodTemplate](https://strimzi.io/docs/operators/latest/full/configuring.html#type-PodTemplate-reference) API is used in many places. So is the ContainerTemplate. So any changes would impact all the component APIs and make it strange if only one actually uses it.